src/coreclr/jit/ssabuilder.cpp

```diff
@@ -1362,6 +1362,71 @@ void Compiler::JitTestCheckSSA()
 }
 #endif // DEBUG
 
+class IncrementalLiveInBuilder
+{
+    Compiler*               m_comp;
+    ArrayStack<BasicBlock*> m_queue;
+
+public:
+    IncrementalLiveInBuilder(Compiler* comp)
+        : m_comp(comp)
+        , m_queue(comp->getAllocator(CMK_SSA))
+    {
+    }
+
+    void MarkLiveInBackwards(unsigned lclNum, const UseDefLocation& use, const UseDefLocation& reachingDef);
+};
+
+//------------------------------------------------------------------------
+// MarkLiveInBackwards: Given a use and its reaching definition, mark that
+// local as live-in into all blocks on the path from the reaching definition to
+// the use.
+//
+// Parameters:
+//   lclNum      - The local
+//   use         - The use
+//   reachingDef - The reaching definition of the use
+//
+void IncrementalLiveInBuilder::MarkLiveInBackwards(unsigned              lclNum,
+                                                   const UseDefLocation& use,
+                                                   const UseDefLocation& reachingDef)
+{
+    if (use.Block == reachingDef.Block)
+    {
+        // No work to be done
+        return;
+    }
+
+    if (!m_comp->AddInsertedSsaLiveIn(use.Block, lclNum))
+    {
+        // We've already marked this block as live-in before -- no need to
+        // repeat that work (everyone should agree on reaching defs)
+        return;
+    }
+
+    m_queue.Reset();
+    m_queue.Push(use.Block);
+
+    while (!m_queue.Empty())
```

Comment:
Maybe add a check that you actually do find the def block, and/or don't  try to search past the first block? (that is, def dominates use)

Reply:
I don't think we necessarily find the reaching def (we might hit a block marked live-in already due to another use), but makes sense to add an assert that we don't get to the first block. I added that assert into `AddInsertedSsaLiveIn` since it never makes sense to mark these added locals as live into the first BB.